src/algebra/category/Group/epi_mono.lean

```diff
@@ -0,0 +1,334 @@
+  nth_rewrite 1 [show (f.range : set B) = 1 *l f.range, from (one_left_coset _).symm],
+  rw [left_coset_eq_iff, mul_one],
+  exact subgroup.inv_mem _ hb,
+end
+
+lemma from_coset_ne_of_nin_range {b : B} (hb : b ∉ f.range) :
+  from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩ ≠
+  from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+begin
+  intros r,
+  simp only [subtype.mk_eq_mk] at r,
+  change b *l f.range = f.range at r,
+  nth_rewrite 1 [show (f.range : set B) = 1 *l f.range, from (one_left_coset _).symm] at r,
+  rw [left_coset_eq_iff, mul_one] at r,
+  exact hb (inv_inv b ▸ (subgroup.inv_mem _ r)),
+end
+
+
+noncomputable instance : decidable_eq X' := classical.dec_eq _
+
+/--
+Let `τ` be the permutation on `X'` exchanging `f.range` and the point at infinity.
+-/
+noncomputable def tau : SX' := equiv.swap
+(from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) ∞
+
+local notation `τ` := tau f
+
+lemma τ_apply_infinity :
+  τ ∞ = from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+equiv.swap_apply_right _ _
+
+lemma τ_apply_from_coset :
+  τ (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) = ∞ :=
+equiv.swap_apply_left _ _
+
+lemma τ_apply_from_coset' (x : B) (hx : x ∈ f.range) :
+  τ (from_coset ⟨x *l f.range.carrier, ⟨x, rfl⟩⟩) = ∞ :=
+(from_coset_eq_of_mem_range _ hx).symm ▸ τ_apply_from_coset _
+
+lemma τ_symm_apply_from_coset :
+  (equiv.symm τ) (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) = ∞ :=
+by rw [tau, equiv.symm_swap, equiv.swap_apply_left]
+
+lemma τ_symm_apply_infinity :
+  (equiv.symm τ) ∞ = from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+by rw [tau, equiv.symm_swap, equiv.swap_apply_right]
+
+/--
+Let `g : B ⟶ S(X')` be defined as as such that, for any `β : B`, `g(β)` is the function sending
+point at infinity to point at infinity and sending coset `y` to `β *l y`.
+-/
+def G : B →* SX' :=
+{ to_fun := λ β,
+  { to_fun := λ x, β • x,
+    inv_fun := λ x, β⁻¹ • x,
+    left_inv := λ x, by dsimp only; rw [←mul_smul, mul_left_inv, one_smul],
+    right_inv := λ x, by dsimp only; rw [←mul_smul, mul_right_inv, one_smul] },
+  map_one' := by ext; simp [one_smul],
+  map_mul' := λ b1 b2, by ext; simp [mul_smul] }
+
+local notation `g` := G f
+
+/--
+Define `h : B ⟶ S(X')` to be `τ g τ⁻¹`
+-/
+noncomputable def H : B →* SX':=
+{ to_fun := λ β, ((τ).symm.trans (g β)).trans τ,
+  map_one' := by ext; simp,
+  map_mul' := λ b1 b2, by ext; simp }
+
+local notation `h` := H f
+
+/--
+The strategy is the following: assuming `epi f`
+* prove that `f.range = {x | h x = g x}`;
+* thus `f ≫ h = f ≫ g` so that `h = g`;
+* but if `f` is not surjective, then some `x ∉ f.range`, then `h x ≠ g x` at the coset `f.range`.
+-/
+
+lemma g_apply_from_coset (x : B) (y : X) :
+  (g x) (from_coset y) =
+  from_coset ⟨x *l y, begin
+    rw [←subtype.val_eq_coe, ←y.2.some_spec, set.mem_range, left_coset_assoc],
+    use x * y.2.some,
+  end⟩ := rfl
+
+lemma g_apply_infinity (x : B) :
+  (g x) ∞ = ∞ := rfl
```

Comment:
Preferred style is, at your option, either
```suggestion
lemma g_apply_infinity (x : B) : (g x) ∞ = ∞ := rfl
```
or
```suggestion
lemma g_apply_infinity (x : B) : (g x) ∞ = ∞ :=
rfl
```
Please adapt all statements in this file accordingly.